src/java/org/apache/cassandra/index/sai/disk/RAMStringIndexer.java

```diff
@@ -0,0 +1,163 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.cassandra.index.sai.disk;
+
+import java.nio.ByteBuffer;
+import java.util.NoSuchElementException;
+
+import org.apache.cassandra.utils.bytecomparable.ByteComparable;
+import org.apache.cassandra.utils.bytecomparable.ByteSource;
+import org.apache.lucene.util.ArrayUtil;
+import org.apache.lucene.util.ByteBlockPool;
+import org.apache.lucene.util.BytesRef;
+import org.apache.lucene.util.BytesRefHash;
+import org.apache.lucene.util.Counter;
+
+/**
+ * Indexes strings into an on-heap inverted index to be flushed in an SSTable attached index later.
+ * For flushing use the PostingTerms interface.
+ */
+public class RAMStringIndexer
+{
+    private final BytesRefHash termsHash;
+    private final RAMPostingSlices slices;
+    private final Counter bytesUsed;
+    
+    private int rowCount = 0;
+    private int[] lastSegmentRowID = new int[RAMPostingSlices.DEFAULT_TERM_DICT_SIZE];
+
+    public RAMStringIndexer()
+    {
+        bytesUsed = Counter.newCounter();
+
+        ByteBlockPool termsPool = new ByteBlockPool(new ByteBlockPool.DirectTrackingAllocator(bytesUsed));
+
+        termsHash = new BytesRefHash(termsPool);
+
+        slices = new RAMPostingSlices(bytesUsed);
+    }
+
+    public long estimatedBytesUsed()
+    {
+        return bytesUsed.get();
+    }
+
+    public boolean isEmpty()
+    {
+        return rowCount == 0;
+    }
+
+    /**
+     * EXPENSIVE OPERATION due to sorting the terms, only call once.
+     */
+    // TODO: assert or throw and exception if getTermsWithPostings is called > 1
+    public TermsIterator getTermsWithPostings()
+    {
+        final int[] sortedTermIDs = termsHash.sort();
+
+        final int valueCount = termsHash.size();
+        final ByteSliceReader sliceReader = new ByteSliceReader();
+
+        return new TermsIterator()
+        {
+            private int ordUpto = 0;
```

Comment:
nit: Maybe just call this `ordinal` or `position` and throw in a quick field JavaDoc indicating this is the position of the current term in the sorted terms array?